### PR TITLE
Test DSL for target system

### PR DESCRIPTION
**Pull request by co0ll3r**
#91 #116
WIP: 

- [x] - optimizeAccum
- [x] - constantFolding
- [x] - resolveDeadlock
- [x] - runSynthesis/Testbench for manual synthesis

### CODE REVIEW

File: test/NITTA/Model/ProcessorUnits/Tests/DSL.hs [hunk 1]
```diff
@@ -215,6 +268,76 @@ setValue var val = do
...
+
+setNetwork network = do
+    st@UnitTestState{unit = ts@TargetSynthesis{}} <- get
+    put st{unit = ts{tMicroArch = network}}
+
+-- | Allows manual binding function. Incompatible with auto synthesis.
+bindInit = do
+    st@UnitTestState{unit = ts@TargetSynthesis{tMicroArch, tDFG}} <- get
+    root <- lift $ getTreeUnit tMicroArch tDFG
+    put st{unit = ts{tMicroArch = root}}
```
[the remark this answers is not in the document]
Comment (ryukzak) [reply test/NITTA/Model/ProcessorUnits/Tests/DSL.hs:302]: Here is a simplified version of this function. 
``` haskell
bindInit = do
    st@UnitTestState{unit = ts@TargetSynthesis{tMicroArch, tDFG}} <- get
    put st{unit = ts{tMicroArch = foldl (flip bind) tMicroArch $ functions tDFG}}
```

So, as you can see, it is possible to bind functions to network on the fly, so you can move it to `assign*` functions.

File: test/NITTA/Model/ProcessorUnits/Tests/DSL.hs [hunk 2]
```diff
@@ -351,26 +311,253 @@ assertCoSimulation =
...
+    , receivedValues :: [(v, [x])]
+    , -- | TODO: add suitable type
+      busType :: Maybe (Proxy x)
+    , report :: Either String (TestbenchReport v x)
+    }
+    deriving (Show)
+
+type Statement u v x r = HasCallStack => StateT (UnitTestState u v x) IO r
+
+type DSLStatement pu v x t r = (HasCallStack, ProcessorUnit pu v x t, EndpointProblem pu v t) => StateT (UnitTestState pu v x) IO r
```
Comment (ryukzak) [inline test/NITTA/Model/ProcessorUnits/Tests/DSL.hs:493]: Rename to PU statement

File: test/NITTA/Model/ProcessorUnits/Tests/DSL.hs [hunk 3]
```diff
@@ -351,26 +311,253 @@ assertCoSimulation =
...
 traceProcess = do
     UnitTestState{unit} <- get
     lift $ putStrLn $ "Process: " <> show (pretty $ process unit)
-    return ()
+
+traceDataflow = do
+    UnitTestState{unit = TargetSystem{mUnit}} <- get
+    lift $ do
+        putStrLn "Dataflow:"
+        mapM_ (\ep -> putStrLn $ "- " <> show ep) $ dataflowOptions mUnit
```
Comment (ryukzak) [inline test/NITTA/Model/ProcessorUnits/Tests/DSL.hs:546]: Extract such prints in separate function. 

See above.

File: test/NITTA/Model/ProcessorUnits/Tests/DSL.hs [hunk 4]
```diff
@@ -50,134 +49,103 @@ unitTestCase "multiplier smoke test" pu $ do     -- 1. Created test case for pro
...
+import NITTA.Synthesis
 import NITTA.Utils
 import Numeric.Interval.NonEmpty hiding (elem)
 import Prettyprinter (pretty)
+import System.Directory
+import System.FilePath
 import Test.Tasty (TestTree)
 import Test.Tasty.HUnit (assertBool, assertFailure, testCase)
 
-unitTestCase ::
```
Comment (ryukzak) [inline test/NITTA/Model/ProcessorUnits/Tests/DSL.hs:151]: Put it back

File: test/NITTA/Tests.hs [hunk 5]
```diff
@@ -24,36 +25,39 @@ module NITTA.Tests (
...
 import NITTA.Model.ProcessorUnits
-import NITTA.Model.Tests.Internals
+import NITTA.Model.ProcessorUnits.Tests.Providers
 import NITTA.Model.Tests.Providers
-import NITTA.Synthesis
 import Test.Tasty (TestTree, testGroup)
 import Test.Tasty.HUnit
 import Test.Tasty.TH
 
 -- FIXME: avoid NITTA.Model.Tests.Internals usage
```
Comment (BasicEC) [inline test/NITTA/Tests.hs:44]: I guess this line should be deleted, since this tests rewrited with DSL

**Review by BasicEC — APPROVED** [review]
Looks cool, as far as I understand I will need to add `doBindPU` when it's merged

**Review by ryukzak — APPROVED** [review]

**Pull request merged**